Quota: Do not count folders towards dashboard quota

conorevans opened the pull request:
Signed-off-by: Conor Evans <coevans@tcd.ie>

**What this PR does / why we need it**: This PR aims to introduce functionality that will allow for QuotaScope to be extended with any relevant conditions beyond just the target model. It's needed for the following, per @marefr : 

> Talked to Carl about it and it's not super important, but would be nice to fix because it creates some support burden.

**Which issue(s) this PR fixes**: 

Fixes #31317

**Special notes for your reviewer**:

Hi @marefr I will tag you here as this is a Draft/WIP. I wanted to solicit your feedback before going any further // implementing tests, lest this be an avenue that you do not think works. I will leave some comments at various points in the PR for your judgment.

A lot of this PR is simply `go fmt`, which is (somewhat) unavoidable.

conorevans committed:
Introduce TargetConditions for QuotaScope - allowing for dashboards that are folders to not count towards quota

Signed-off-by: Conor Evans <coevans@tcd.ie>

conorevans commented on pkg/services/quota/quota.go:
```diff
@@ -118,40 +128,107 @@ func (qs *QuotaService) getQuotaScopes(target string) ([]models.QuotaScope, erro
+
+// get any extra conditions that are required to properly
+// determine the quota for a given target
+func (qs *QuotaService) getQuotaScopeConditions(target string) string {
```
The reason I chose here for the function was that I preferred it be passed up the chain to the Query. Additionally this centralizes the conditions based on the target, rather than having random SQL conditions scattered around various models (well, potentially scattered, since this may be the only occasion, or one of few, where additional conditions are required in order to identify the correct quota).

conorevans replied:
I'm also open to the name, since it's conditions related to the target of the quota scope rather than the quota scope itself.  I considered `getQuotaScopeConditionsByTarget` but I felt like that may be redundant, and isn't done above (i.e. it is `QuotaReached` and not `QuotaReachedByTarget`)

conorevans commented on pkg/services/sqlstore/quota.go:
```diff
@@ -231,7 +235,9 @@ func UpdateUserQuota(cmd *models.UpdateUserQuotaCmd) error {
-	rawSQL := fmt.Sprintf("SELECT COUNT(*) as count from %s", dialect.Quote(query.Target))
+	rawSQL := fmt.Sprintf("SELECT COUNT(*) AS count FROM %s %s",
+		dialect.Quote(query.Target),
+		query.TargetConditions)
```
ah this would not work here because the global quota does not use any `WHERE`s beforehand (unlike above with `WHERE user_id=?` or `WHERE org_id=?`) so it'd be `SELECT COUNT(*) AS count FROM dashboard AND is_folder = 0`.

Hmm. I suppose the cleanest way to do it would be to just set the conditions (i.e. `is_folder = 0`)  for the QuotaScope (and eventually the `Get*QuotaByTargetQuery`), and then in each place insert this condition with the appropriate operators.

So here it would be something like:

```
if query.TargetConditons != "" {
  targetConditions := fmt.Sprintf("WHERE %s", query.TargetConditions)
} else {
  targetConditions := ""
}

rawSQL := fmt.Sprintf("SELECT COUNT(*) AS count FROM %s %s", dialect.Quote(query.Target), targetConditions)
```

conorevans replied:
Whereas above (e.g. for `GetOrgQuotaByTarget`) it would be:

```
if query.TargetConditons != "" {
  targetConditions := fmt.Sprintf("AND %s", query.TargetConditions)
} else {
  targetConditions := ""
}

rawSQL := fmt.Sprintf("SELECT COUNT(*) AS count FROM %s WHERE org_id=? %s", dialect.Quote(query.Target), targetConditions)
```

conorevans replied:
I don't love it, but I don't hate it. Let me know if you see a better approach.

conorevans replied:
Oh actually, I could pass the models.ReqContext at 

```
func (qs *QuotaService) getQuotaScopeConditions(target string) string {
```

and return the appropriate global vs non-global string there.

marefr replied:
Not sure. One way would be to add `WHERE 1=1` :smile:

conorevans replied:
Haha I did consider that but I figured `pr/external` wouldn't get that by 😉.

conorevans replied:
Though honestly, it could be cleaner to have that in the query than having the `WHERE` quite a way downstream. Unless I can think of a nicer approach, I will take this approach - you or another reviewer can then chime in if you think of a nicer approach.

conorevans commented on pkg/models/quotas.go:
```diff
@@ -18,9 +18,10 @@ type Quota struct {
-	DefaultLimit int64
+	Name             string
+	Target           string
+	TargetConditions string
```
I considered making this `[]string` such that each condition (if there were multiple) would be more clear to discern.

conorevans replied:
I think that could be equally as messy as it would be somewhat clearer to discern, however. It would also make https://github.com/grafana/grafana/pull/32519#discussion_r604681078 unrealistic as an approach - if that approach is adjudged to be appropriate.

marefr replied:
This should be fine. Would that be needed in the future we can change it

marefr commented on pkg/models/quotas.go:
```diff
@@ -56,10 +58,11 @@ type GetOrgQuotasQuery struct {
-	Default int64
-	Result  *UserQuotaDTO
+	Target           string
+	TargetConditions string
```
Think you can skip adding `TargetConditions` for `GetUserQuotaByTargetQuery`. Can be added in the future, if needed

conorevans replied:
Yeah, wasn't sure on whether to do this or not. Sounds good to me.

marefr commented on pkg/services/quota/quota.go:
```diff
@@ -93,7 +98,12 @@ func (qs *QuotaService) QuotaReached(c *models.ReqContext, target string) (bool,
 				continue
 			}
-			query := models.GetUserQuotaByTargetQuery{UserId: c.UserId, Target: scope.Target, Default: scope.DefaultLimit}
+			query := models.GetUserQuotaByTargetQuery{
```
This change can be skipped based on my earlier comment

marefr commented on pkg/services/quota/quota.go:
```diff
@@ -118,40 +128,107 @@ func (qs *QuotaService) getQuotaScopes(target string) ([]models.QuotaScope, erro
 		scopes = append(scopes,
-			models.QuotaScope{Name: "global", Target: target, DefaultLimit: qs.Cfg.Quota.Global.User},
-			models.QuotaScope{Name: "org", Target: "org_user", DefaultLimit: qs.Cfg.Quota.Org.User},
+			models.QuotaScope{
+				Name:             "global",
+				Target:           target,
+				TargetConditions: qs.getQuotaScopeConditions(target),
+				DefaultLimit:     qs.Cfg.Quota.Global.User,
+			},
+			models.QuotaScope{
+				Name:             "org",
+				Target:           "org_user",
+				TargetConditions: qs.getQuotaScopeConditions(target),
+				DefaultLimit:     qs.Cfg.Quota.Org.User,
+			},
```
Think we can skip this for now since not needed. TargetConditions will be empty by default

marefr commented on pkg/services/quota/quota.go:
```diff
@@ -118,40 +128,107 @@ func (qs *QuotaService) getQuotaScopes(target string) ([]models.QuotaScope, erro
 		scopes = append(scopes,
-			models.QuotaScope{Name: "global", Target: target, DefaultLimit: qs.Cfg.Quota.Global.Org},
-			models.QuotaScope{Name: "user", Target: "org_user", DefaultLimit: qs.Cfg.Quota.User.Org},
+			models.QuotaScope{
+				Name:             "global",
+				Target:           target,
+				TargetConditions: qs.getQuotaScopeConditions(target),
+				DefaultLimit:     qs.Cfg.Quota.Global.Org,
+			},
+			models.QuotaScope{
+				Name:             "user",
+				Target:           "org_user",
+				TargetConditions: qs.getQuotaScopeConditions(target),
+				DefaultLimit:     qs.Cfg.Quota.User.Org,
+			},
```
Think we can skip this for now since not needed. TargetConditions will be empty by default

marefr commented on pkg/services/quota/quota.go:
```diff
@@ -118,40 +128,107 @@ func (qs *QuotaService) getQuotaScopes(target string) ([]models.QuotaScope, erro
 		scopes = append(scopes,
-			models.QuotaScope{Name: "global", Target: target, DefaultLimit: qs.Cfg.Quota.Global.DataSource},
-			models.QuotaScope{Name: "org", Target: target, DefaultLimit: qs.Cfg.Quota.Org.DataSource},
+			models.QuotaScope{
+				Name:             "global",
+				Target:           target,
+				TargetConditions: qs.getQuotaScopeConditions(target),
+				DefaultLimit:     qs.Cfg.Quota.Global.DataSource,
+			},
+			models.QuotaScope{
+				Name:             "org",
+				Target:           target,
+				TargetConditions: qs.getQuotaScopeConditions(target),
+				DefaultLimit:     qs.Cfg.Quota.Org.DataSource,
+			},
```
Think we can skip this for now since not needed. TargetConditions will be empty by default

marefr commented on pkg/services/quota/quota.go:
```diff
@@ -118,40 +128,107 @@ func (qs *QuotaService) getQuotaScopes(target string) ([]models.QuotaScope, erro
 		scopes = append(scopes,
-			models.QuotaScope{Name: "global", Target: target, DefaultLimit: qs.Cfg.Quota.Global.ApiKey},
-			models.QuotaScope{Name: "org", Target: target, DefaultLimit: qs.Cfg.Quota.Org.ApiKey},
+			models.QuotaScope{
+				Name:             "global",
+				Target:           target,
+				TargetConditions: qs.getQuotaScopeConditions(target),
+				DefaultLimit:     qs.Cfg.Quota.Global.ApiKey,
+			},
+			models.QuotaScope{
+				Name:             "org",
+				Target:           target,
+				TargetConditions: qs.getQuotaScopeConditions(target),
+				DefaultLimit:     qs.Cfg.Quota.Org.ApiKey,
+			},
```
Think we can skip this for now since not needed. TargetConditions will be empty by default

marefr commented on pkg/services/quota/quota.go:
```diff
@@ -118,40 +128,107 @@ func (qs *QuotaService) getQuotaScopes(target string) ([]models.QuotaScope, erro
 	case "session":
 		scopes = append(scopes,
-			models.QuotaScope{Name: "global", Target: target, DefaultLimit: qs.Cfg.Quota.Global.Session},
+			models.QuotaScope{
+				Name:             "global",
+				Target:           target,
+				TargetConditions: qs.getQuotaScopeConditions(target),
+				DefaultLimit:     qs.Cfg.Quota.Global.Session,
+			},
```
Think we can skip this for now since not needed. TargetConditions will be empty by default

marefr commented on pkg/services/sqlstore/quota.go:
```diff
@@ -139,7 +141,9 @@ func GetUserQuotaByTarget(query *models.GetUserQuotaByTargetQuery) error {
 
 	// get quota used.
-	rawSQL := fmt.Sprintf("SELECT COUNT(*) as count from %s where user_id=?", dialect.Quote(query.Target))
+	rawSQL := fmt.Sprintf("SELECT COUNT(*) AS count FROM %s WHERE user_id=? %s",
+		dialect.Quote(query.Target),
+		query.TargetConditions)
```
Think we can skip this for now since not needed.

marefr reviewed (commented):
Added a couple comments. Looks good in general, but think this PR can be simplified by removing some changes not needed.

conorevans commented:
> Btw, there's some tests in sqlstore/quota_test.go that would be nice to update/add cover of the functionality you've added for dashboards

For sure, wanted to get initial feedback on approach before implementing potentially changing tests. Will add appropriate tests soon.

conorevans committed:
PR review: remove TargetConditions from GetUserQuotaByTargetQuery, remove unnecessary additions of TargetConditions since default is already set to "", include WHERE 1=1 in global query to ensure that TargetConditions is uniform irrespective of context

Signed-off-by: Conor Evans <coevans@tcd.ie>

conorevans commented:
Hi @marefr I addressed your comments. I do think the `WHERE 1=1` is actually cleaner than differing conditional logic downstream, so I went ahead with it. If you or the other reviewers have other ideas, I'm open to them.

With respect to:

> Btw, there's some tests in sqlstore/quota_test.go that would be nice to update/add cover of the functionality you've added for dashboards.

Now while I have coded a few microservices (nanoservices, really), I haven't dealt with a testing suite as large as this before. I delved into the test file to understand how to go about it, but as far as I can see, the tests in `quotas_test.go` are relatively naive, just setting quotas to 5, the initial count to 4, then one successful test + increment of current count, then one test that fails since the count == quota. The requests are all mocked based on model, but don't include any actual model mocks wherein the parameters would be set/mocked (`is_folder` in this case).

I think the test will have to be added to `dashboard_test` under `postDashboardScenario` by setting `setting.Quota` to be something like:

```go
		setting.Quota = setting.QuotaSettings{
			Enabled: true,
			Org: &setting.OrgQuota{Dashboard:  X}
			Global: &setting.GlobalQuota{Dashboard:  X},
		}
```

and then testing a post where X dashboards exist and the `X+1`th has `is_folder = 1` so isn't counted?

Does that sounds right to you? Or is there testing capabilities to achieve what we want in `sql_store/quota_test.go` that just aren't currently exposed because no model has thus far had `TargetConditions`.

Thanks!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=32519) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=32519) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=32519) it.</sub>

ethirolle commented:
@marefr do we have an ETA for this fix? We have some GC customers recently hitting limits due to folders being included in the dashboard count. Thanks!!

marefr commented:
I haven't had time to come back to this one in a while. Sorry for that. I'm planning to get this merged soon.

conorevans committed:
merge master, resolve conflicts

Signed-off-by: Conor Evans <coevans@tcd.ie>

conorevans commented:
Resolved conflicts so hopefully that should help smooth the process for this PR, still have to add tests ofc but waiting on guidance from above (feel free to hijack the branch if it's quicker for you, i won't mind)

conorevans committed:
missed a comma when resolving conflicts

Signed-off-by: Conor Evans <coevans@tcd.ie>

marefr committed:
Merge remote-tracking branch 'origin/main' into conorevans/exclude-folders-from-dashboard-quota

marefr committed:
Simplify and refactor to standard lib for tests

marefr committed:
Merge remote-tracking branch 'origin/main' into conorevans/exclude-folders-from-dashboard-quota

marefr commented:
I decided to simplify this and added the condition in the sqlstore package. Main reason is because boolean conditions differ between sqlite/mysql and postgres and the only target in need of additional conditions is currently dashboard.

marefr reviewed (approved):
LGTM

marefr commented:
@conorevans thank you for contributing to Grafana! Sorry it took some time for us to get this merged.
